module/zfs/dnode.c

```diff
@@ -1778,15 +1778,23 @@ dnode_try_claim(objset_t *os, uint64_t object, int slots)
 }
 
 /*
- * Checks if the dnode contains any uncommitted dirty records.
+ * Checks if the dnode itself is dirty, or is carrying any uncommitted records.
+ * It is important to check both conditions, as some operations (eg appending
+ * to a file) can dirty both as a single logical unit, but they are not synced
+ * out atomically, so checking one and not the other can result in an object
+ * appearing to be clean mid-way through a commit.
+ *
+ * Do not change this lightly! If you get it wrong, dmu_offset_next() can
+ * detect a hole where there is really data, leading to silent corruption.
  */
 boolean_t
 dnode_is_dirty(dnode_t *dn)
 {
 	mutex_enter(&dn->dn_mtx);
 
 	for (int i = 0; i < TXG_SIZE; i++) {
-		if (multilist_link_active(&dn->dn_dirty_link[i])) {
+		if (multilist_link_active(&dn->dn_dirty_link[i]) ||
+		    list_head(&dn->dn_dirty_records[i])) {
```

Comment:
```suggestion
		    !list_is_empty(&dn->dn_dirty_records[i])) {
```

Reply:
Made the code change. I've made the commit message a little more passive so as not to assert any particular thing, just the observation, because I'm still not sure where the gap actually comes from (I've two or three ideas). This shape of fix is definitely the right one regardless; we can work on understanding it better later.

Thanks!

Reply:
As I see it, dmu_objset_sync_dnodes() first calls multilist_sublist_remove(), cleaning dn_dirty_link, and only then dnode_sync() to actually work on the dirty records.

Reply:
@amotin I never did thank you for this. I had assumed that somewhere `dn_dirty_link` wasn't being set, or set too late. I hadn't considered it was being cleared too early, and from there the whole thing opened up. I don't know if you properly knew or just had an intuition about it. Regardless, it was the bit of info I needed. Thank you!